### ISSUE TITLE
Download VODs, highlights and uploads

### ISSUE THREAD

**Issue by HugoJF** [HugoJF/twitch-clip-downloader#13]
Since lots of streamers are trying to download VODs and stuff before deleting them to avoid DMCA, I've realized that this tool could be easily updated to also support VOD, highlights and upload downloads.

This would take a bit of work but should be pretty straight-forward.

If you think VOD downloading could be useful, please +1 this issue as a feedback.

**Comment by HugoJF** [comment]
This has been partially implemented in commit 74313ef10a104fb5749c44ddb5fec14e91f3f708.

For whatever reason, youtube-dl is not downloading any videos (clips are working fine and they are sharing the same code).

When youtube-dl is initialized in NodeJS and the `info` event is called, `info._filename` is undefined. Attempting to download the same URL from youtube-dl directly works, which makes me wonder how `node-youtube-dl` wraps youtube-dl.

I'll resume working on this on a later date, if anyone know what is happening, let me know!

**Comment by HugoJF** [comment]
Even though the `vods` branch works, `youtube-dl` doesn't download fragments concurrently and spawning multiple instances use alot of resources, I decided to re-write the VOD downloader part without `youtube-dl` and with that, almost rewrite the entire tool to fix alot of bad practices (since this tool wasn't meant to become what it became).

Today is the second day I worked on it and pretty much got fragment downloads (and almost everything else rewritten with classes) done. Since `youtube-dl` is not responsible for downloads, the tool is (finally) able to saturate my 300Mbps line (with 50 parallel downloads while only using 7% of my 4790k) both for clips and VODs.

I expect (and hope) the whole rewrite to be done this week.
